components/places/src/bookmark_sync/engine.rs

```diff
@@ -490,14 +491,42 @@ fn apply_remote_items(db: &PlacesDb, scope: &SqlInterruptScope, now: Timestamp)
         }
     ).ok(); // We ignore the error because it's expected, users should not have entries with the same id that have different types
 
+    // Same as above but we check if any users have any undetected orphaned bookmarks and
+    // report them appropiately
+    let orphaned_count: i64 = db.query_row(
+        "WITH RECURSIVE orphans(id) AS (
+           SELECT b.id
+           FROM moz_bookmarks b
+           WHERE b.parent IS NOT NULL
+             AND NOT EXISTS (
+               SELECT 1 FROM moz_bookmarks p WHERE p.id = b.parent
+             )
+           UNION
+           SELECT c.id
+           FROM moz_bookmarks c
+           JOIN orphans o ON c.parent = o.id
+         )
+         SELECT COUNT(*) FROM orphans;",
+        [],
+        |row| row.get(0),
+    )?;
+
+    if orphaned_count > 0 {
```

Comment:
We also talked a little about doing this both before (like here) and after (somewhere below) a sync so we can get a better idea whether dogear really does fix all cases up - I think this would be useful to know because it might inform next steps.

(I'm also very very slightly worried about perf here for users with many thousands of deep bookmarks. I wonder if it might be worth counting bookmarks first and only doing these expensive recursive checks if there are less than, say 500. Really happy for you to decide this though!)

Reply:
Great idea! I decided to add a breadcrumb here instead that will be added to the report later if dogear ends up _not_ fixing it. This will prevent any noise/red herrings that were orphans here but fixed up later.

I wrote a quick test on testing that query with 9k bookmarks (but not deeply nested) and it ran about 7ms. Which doesn't tell us much on a real android or iOS device but since this is something we'll be closely watching in nightly, I'd like to cast as wide a net as possible here and we can quickly limit/eliminate this depending on what the data will tell us in the next few days